File: system-test/dotnet/Driver.Adapter/External/Erp/ErpStubDriver.cs
using Common;
using Driver.Port.External.Erp.Dtos;
using Driver.Port.External.Erp.Dtos.Error;
using Driver.Adapter.External.Erp.Client;
using Driver.Adapter.External.Erp.Client.Dtos;
using Driver.Adapter.External.Erp.Client.Dtos.Error;

namespace Driver.Adapter.External.Erp;

/// <summary>
/// ErpStubDriver uses WireMock to stub ERP API responses.
/// This allows tests to run without a real ERP system.
/// </summary>
public class ErpStubDriver : BaseErpDriver<ErpStubClient>
{
    public ErpStubDriver(string baseUrl) : base(new ErpStubClient(baseUrl))
    {
    }

    public override async ValueTask DisposeAsync()
    {
        await _client.RemoveStubsAsync();
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    public override Task<Result<VoidValue, ErpErrorResponse>> ReturnsProductAsync(ReturnsProductRequest request)
    {
        var extProductDetailsResponse = new ExtProductDetailsResponse
        {
            Id = request.Sku!,
            Title = string.Empty,
            Description = string.Empty,
            Price = decimal.Parse(request.Price!),
            Category = string.Empty,
            Brand = string.Empty
        };

        return _client.ConfigureGetProductAsync(extProductDetailsResponse)
            .MapErrorAsync(error => MapError(error));
    }

    public override Task<Result<VoidValue, ErpErrorResponse>> ReturnsPromotionAsync(ReturnsPromotionRequest request)
    {
        var extGetPromotionResponse = new ExtGetPromotionResponse
        {
            PromotionActive = request.PromotionActive,
            Discount = decimal.Parse(request.Discount!)
        };

        return _client.ConfigureGetPromotionAsync(extGetPromotionResponse)
            .MapErrorAsync(error => MapError(error));
    }
}

File: system-test/java/src/main/java/com/optivem/shop/dsl/driver/adapter/external/erp/ErpStubDriver.java
package com.optivem.shop.dsl.driver.adapter.external.erp;

import com.optivem.shop.dsl.driver.adapter.external.erp.client.ErpStubClient;
import com.optivem.shop.dsl.driver.adapter.external.erp.client.dtos.ExtProductDetailsResponse;
import com.optivem.shop.dsl.driver.port.external.erp.dtos.ReturnsProductRequest;
import com.optivem.shop.dsl.driver.port.shared.dtos.ErrorResponse;
import com.optivem.shop.dsl.common.Converter;
import com.optivem.shop.dsl.common.Result;

/**
 * ErpStubDriver uses WireMock to stub ERP API responses.
 * This allows tests to run without a real ERP system.
 */
public class ErpStubDriver extends BaseErpDriver<ErpStubClient> {
    public ErpStubDriver(String baseUrl) {
        super(new ErpStubClient(baseUrl));
    }

    @Override
    public Result<Void, ErrorResponse> returnsProduct(ReturnsProductRequest request) {
        var extProductDetailsResponse = ExtProductDetailsResponse.builder()
                .id(request.getSku())
                .title("")
                .description("")
                .price(Converter.toBigDecimal(request.getPrice()))
                .category("")
                .brand("")
                .build();

        return client.configureGetProduct(extProductDetailsResponse)
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }
}
